### ISSUE TITLE
Error Importing MBOX files with Subdirectory Option

### ISSUE THREAD

**Issue by TalismanicTortoise2468** [thundernest/import-export-tools-ng#41]
Hi there! 

First, thank you for your work on this extension, this tool is exceptionally useful and your work to maintain it helps A LOT of people out.

As for the issue, it seems that there is a problem importing MBOX files using the "_Select a directory where searching the mbox files to import (also in subdirectories)_" option. I assume that this option allows the user to select a directory where the extension will search in (as well as it's subdirectories) for MBOX files to import, similar to the "_also from its subdirectories_" option when importing EML files.

However, it currently appears that this is not the case. Unlike the EML option, there is no notification at the bottom telling the user that the directory is being traversed and updating on the status of the import. In my tests it seems to not be attempting to import any files.

I have no problem importing the same MBOX files directly or with the 
"_Select a directory where searching the mbox files to import_" option, so I don't believe that it has to do with the files which I am using. 

I am unfamiliar with the development of Thunderbird extensions or I would have looked through the repository more thoroughly to try to offer more narrowed guidance. 

Thank you in advance!

SPECS:
ImportExportToolsNG Version - 4.0.4
Thunderbird Version - 68.2.1
OS - Windows 10

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
Thanks for the issue post and the support!  As you can tell I have "adopted" this teeth numbered extension as the Thunderbird extension developer liaison.  I have to always state upfront , I still have plenty to learn about this complex extension.  There are several things in the menu that don't have the best explanations.  If I can finally get everything upgraded I'm responsible for my next step is to make a small user guide and improve the menus.  In the meantime I am still discovering new edge cases.

mbox import is probably second only to EM L imports.  I just went through a similar import issue from a user that had exported from Apple Mail.  I just did import of that which has three folders top two subfolders.  While they import, the importance structure depends on some naming or set up conventions.  I have not fully figured out all the permutations.  I should be able to do some testing time in tomorrow to see if I can make a test and example.

Question:  can you give me an example of your naming/folder structure.  I want to make a test that matches your set up as much as possible.  
cheers
@cleidigh

**Comment by TalismanicTortoise2468** [comment]
@cleidigh 
As in the naming of the folder where I am trying to import? I'm not sure I can do that exactly, but I can probably rename them, test it, and then provide you with that. Just send me a list of everything that you need and I'll get it to you soonish.
@TalismanicTortoise2468

**Comment by cameronjpu** [comment]
Just FYI, I am trying to import data from WLM and I am having the same issue.  The output from the export is the folder structure I'm seeking but zero emails inside the folders.  Thanks for your time!

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
@cameronjpu 
Thanks.  I have been doing various experiments because this not entirely obvious.
I just went through doing some support on a WLM issue that really was not an issue so I am surprised about an issue there.  can you give any more details?  I assume you're importing EMl files?
 
Concerning structure with mboxes my testing so far shows that multiple nested mailboxes get imported in a flat structure.  I am still trying to figure out if this was always the case or something has changed.
@TalismanicTortoise2468 I don't want to ask you for anything yet, I am able to test myself at this point.


Stay tuned.

**Comment by cleidigh** [comment]
@cameronjpu 
Since I believe you are trying to import EML files I would like to open separate issue for tracking.
just confirm if I have anything incorrect.

**Comment by TalismanicTortoise2468** [comment]
@cleidigh 
Thanks for all this! I would like to clarify my issue based on the above. 

I am interested in importing all the MBOX files in a computer's directory as well as all MBOX files of that directory's subdirectories. My issue is not with the folders identified within the MBOX file being imported, but rather the inability to import multiple MBOX files from a tree of directories on the computer. 

I just wanted to ensure that I articulated what my issue was as I recognized that talking about folders can lead to miscommunication on terminology. Thanks!!

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
just saw your post, hoping to do a bit of back-and-forth .  there are so many permutations clarity as you to his very important.

**Comment by TalismanicTortoise2468** [comment]
@cleidigh 
Just let me know how I can help

**Comment by cleidigh** [comment]
your clarifications are good, the problem is clear : the need to import a set of e-mail folders that have nested subfolders.  I have been testing various structures to see how they are actually imported.
I am also working on better understanding the differences between the importation of different file types. the extension is not completely consistent in functionality for different file formats.
Here are a couple of points to know. (if I ever catch up I will do the user guide!)
- it is possible to import a set of mbox files either where all files are in one system folder or with nested subfolders with one or more box files within a subfolder.
- the extension attempts to use the filename of the mbox file.  subfolder1.mbox ==> subfolder1.mbox (ThunderbirdFolderName)
- it will add an integer suffix if it finds conflicting names

So one issue I have been struggling with is the separation or distinction of the folder containers versus the actual filenames. 
The other issue directly related to structure is that in my tests I am able to import multiple Mboxes from a folder tree including subfolder files. The consistent result is a flattening meaning that all boxes are imported the same level, clearly not desired or expected result.

Putting aside the naming convention from moment, I am trying to figure out why the structure is removed on import, perhaps to state the obvious. The nonobvious aspects are that there are three levels of file structures involved in the importation.  I believe one of the mappings has an error probably introduced by Thunderbird changes.

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
So taking into account your original post and the above, I understand that you can successfully import multiple lamp boxes but you lose structure.  If that is correct, we observed the same behavior.
In the that case I can hopefully debug on my side without needing you until I have something I believe works.  The issues where I cannot reproduce problem much more difficult. of course we still have to solve this one!  
FYI I am not really fast as I am disabled and have to do all development by voice.  That phone also makes any non-Windows work very very very slow (see my OS X problem)
Anyway let me know if I've anything wrong or if anything I have said inspires more info.

Hopefully the fix straightforward
@cleidigh

**Comment by TalismanicTortoise2468** [comment]
@cleidigh 
Thank you for such a quick reply. That last post helps a lot! No, I am not able to import multiple MBOX files when using the "_Select a directory where searching the mbox files to import (also in subdirectories)_" option. Currently, if I choose that option and then select a folder with the following structure:

TopFolder
|- FirstLevel1.mbox
|- FirstLevel2.mbox
--- SubFolder
    |- SecondLevel.mbox

I observe no discernible change to my local folders in Thunderbird. I would expect to see a new local folder names "TopLevel" and a sub-folder names "SubFolder".

From what I understand, you are able to import the above structure as follows:

TopFolder
|- FirstLevel1.mbox
|- FirstLevel2.mbox
|- SecondLevel.mbox

If that is the case, then that would mean you cannot reproduce my issue, and I can do further testing on my part. Currently, I cannot import MBOX files using the above structure and the "_Select a directory where searching the mbox files to import (also in subdirectories)_" option.

Thanks again for all the help!
@TalismanicTortoise2468

**Comment by TalismanicTortoise2468** [comment]
Markdown messed up my first ASCII chart. I think this clarifies:

TopFolder
|- FirstLevel1.mbox
|- FirstLevel2.mbox
--- SubFolder
--- |- SecondLevel.mbox

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
So no e-mails imported, obviously no structure either.
I am still trying to understand 'what importation successful?

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
Also I don't think I have requested the error console log (Shift-Control-J)  ThenDoTheOperations and copy the entire window to post here.  Usually with a clear bug situation like you have there will be something to go on.

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
okay lots of testing and better understanding 👍

After a ton of experimentation as well as looking at other sources, I believe I now
have a much better understanding of what I think the extension should do.  There are
certainly several subtle behaviors most of which I reference below.  

Recapping where we are:
- I can use the extension with the same OS et cetera and each of the four options
  appear to behave correctly and as designed.  
- Only option 2 understands and retains structure, everything else is flattened
- mbox files (as the extension understands) contain no structure themselves
- You appear to be able to use (1) and probably (3) importing the same mbox files
- You try to use (4) but nothing gets imported, no folders created within Thunderbird 

As  you pointed out, assuming everything above is correct, we have different results
While I do not believe there is a problem with your files, it's better for us to test the same thing to start
I posted two things tonight. A beta 4.0.5-b1 xpi pool which has more debug output
Also there is a new folder 'test' with a zip file of various test trees.
TopLevel under the MBOX directory matches your example
TopLevelStructured is a slightly simplified tree organized as the extension expects it
for option (2).  

I would suggest trying (2) with TopLevelStructured selecting TopLevelwhen you import
You should be able to import with the other options as well.  We need to get the same
results for that as well as (3) or (4)


For reference/clarity below are the 4 import options for mboxes
The original wording is not as clear as it should be (on my to do list)

1)  Import directly one or more mbox files

2)  Import one or more mbox files, with its/their subdirectory
     Select just the mbox file, the directory with the same name
     and the extension \'sbd\' will be automatically imported, if it exists.

3)  Select a directory where searching the mbox files to import

4)  Select a directory where searching the mbox files to import (also in subdirectories)

Brief description of each options usage:

1)  Import directly one or more mbox files

This is the most basic option to import specific mbox files (using the picker)
First the folder to import to is selected within Thunderbird (should be a local folder)
For each valid mbox file a new subfolder WITH THE NAME OF THE SELECTED FILE (may have suffix number added )
So two files will create two folders each carrying the e-mails of the corresponding mbox
Note: it does not combine mbox files ever as far as I can tell

2)  Import one or more mbox files, with its/their subdirectory

This option is the only one which imports with "structure", meaning it understands
and maintains  the structure defined by the import source tree.
The extension only understands a specific layout of the source tree in order to re-create
it within Thunderbirds folders.  The following input tree represents an example of
a simple tree with a TopLevel folder with one subfolder.

ImportContainer (d)
  TopLevel (mbox file, no extension)
  TopLevel.sbd (d) 
    SubLevel1 (mbox file, no extension)
	SubLevel1 (d) 

To import the above one would first select the import location within Thunderbird
for example "ImportFolder" then after choosing option (2) in the import dialog
use the file to select the top mbox file that has a same name subfolder (.sbd)
The extension will then import everything from the top mbox file and include the subfolder

This is the same structure that would result from an export of a similar tree 
Within Thunderbird  the result would look like the following:

ImportFolder (import folder)
  TopLevel (top imported folder)
   (e-mails if any from TopLevel file)
   SubLevel1 (imported subfolder)
     (e-mails if any from SubLevel1 file)


3)  Select a directory where searching the mbox files to import

This option performs the same general method as (1) however, a container folder
is selected as opposed to files.  ANY files within the SINGLE selected folder
that are detected to be mbox type are imported to the selected folder prior to import
The extension does not care about any file extension, in fact if there is an extension
it becomes part of the new folder name.  All mbox files found are imported under
the import folder , all at the same level.

4)  Select a directory where searching the mbox files to import (also in subdirectories)

This option follows the same behavior as (3), however, it will traverse recursively
all the way into the source tree collecting any mbox files it finds and imports them
into the ImportFolder ALL AT THE SAME LEVEL (NO STRUCTURE FROM SOURCE)

(Importing TopLevel like your tree)
![image](https://user-images.githubusercontent.com/25272315/69489688-09065480-0e4a-11ea-8a47-78a31ef2d2bf.png)
100

TopLevelStructured
![image](https://user-images.githubusercontent.com/25272315/69489692-289d7d00-0e4a-11ea-948f-d59acc08198f.png)

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
Pointers
https://github.com/thundernest/import-export-tools-ng/blob/import-tests/xpi/import-export-tools-ng-4.0.5-b1-tb.xpi

https://github.com/thundernest/import-export-tools-ng/blob/import-tests/tests/import.zip

**Comment by TalismanicTortoise2468** [comment]
@cleidigh 
So after some testing, it appears that I cannot replicate the problem that I was having. It may have ultimately been a space issue as I am working with very large MBOX files and am using a machine with a significantly larger storage capability, but I would recommend that this issue be closed.

I apologize for wasting your time, but thank you so much for being responsive and helpful!
@TalismanicTortoise2468

**Comment by TalismanicTortoise2468** [comment]
@cleidigh 

I just saw your extensive reply from 2 days ago. Thank you so much for dedicating all that effort, it makes me feel slightly more ashamed that it was likely a user error. I will run those tests that you provided, but I would bet that they end up with the same behavior you observed.

I'll keep you updated and thanks again!

@TalismanicTortoise2468

**Comment by cleidigh** [comment]
@TalismanicTortoise2468 
No reason to apologize or be ashamed!  The fact of the matter is I do not have the complete understanding of this complex extension given I took it over.  By really doing some deep testing but I do feel I now understand what the extension will and will not do with mboxes.  The lack of a user guide as well as confusing descriptions does not help.  What I wrote above will go along way towards even a basic user guide.  So I thank you for taking the time to go back and forth.  I am sure you have seen others have similar issues even if they turn out to be questions.

Cheers
@cleidigh
